clang/lib/Sema/Metafunctions.cpp

```diff
@@ -4480,6 +4480,73 @@ bool reflect_result(APValue &Result, Sema &S, EvalFn Evaluator,
   return SetAndSucceed(Result, Arg.Lift(Args[1]->getType()));
 }
 
+bool is_nonstatic_member_function(ValueDecl *FD) {
+  if (!FD) {
+    return false;
+  }
+
+  if (dyn_cast<CXXConstructorDecl>(FD)) {
+    return false;
+  }
+
+  auto *MD = dyn_cast<CXXMethodDecl>(FD);
+  if (!MD) {
+    // might be a pointer to member function
+    QualType QT = FD->getType();
+    // check if the type is a pointer to a member
+    if (const MemberPointerType *MPT = QT->getAs<MemberPointerType>()) {
+      QualType PT = MPT->getPointeeType();
+      // check if the pointee type is a function type
+      if (const FunctionProtoType *FPT = PT->getAs<FunctionProtoType>()) {
+        return true;
+      }
+    }
+  } else {
+    return !MD->isStatic();
+  }
+
+  return false;
+}
+
+CXXMethodDecl *getCXXMethodDeclFromDeclRefExpr(DeclRefExpr *DRE) {
+  ValueDecl *VD = DRE->getDecl();
+
+  if (auto *MD = dyn_cast<CXXMethodDecl>(VD)) {
+    // method declaration
+    return MD;
+  } else {
+    // pointer to non-static method
+    // validation was done in is_nonstatic_member_function
+    auto *VarD = cast<VarDecl>(VD);
+    if (!VarD->isConstexpr()) {
+      // avoid extracting constexpr method declaration
+      // from non-constexpr function pointers
+      return nullptr;
+    }
+
+    Expr *Init = VarD->getInit();
+
+    if (auto *UnaryOp = dyn_cast<UnaryOperator>(Init)) {
+      // get the operand of &
+      Init = UnaryOp->getSubExpr();
+    }
+
+    while (auto *CastExpr = dyn_cast<ImplicitCastExpr>(Init)) {
+      // handle ImplicitCastExpr if it is wrapping the actual expression
+      Init = CastExpr->getSubExpr();
+    }
+
+    if (auto *DREFromUnary = dyn_cast<DeclRefExpr>(Init)) {
+      return dyn_cast<CXXMethodDecl>(DREFromUnary->getDecl());
+    }
+  }
+
+  // if we reach this point
+  // then probably there is a bug in `is_nonstatic_member_function`
+  // or we failed to extract method from pointer
+  llvm_unreachable("failed to get member function from decl ref expression");
+}
```

Comment:
I'm not sure that introspecting the substructure of the `DeclRefExpr` is the best approach here. We could instead evaluate it as a prvalue, obtain an `APValue`, and pull the `CXXMethodDecl` out of there.

One of the kinds that an `APValue` might be is [`MemberPointer`](https://github.com/bloomberg/clang-p2996/blob/p2996/clang/include/clang/AST/APValue.h#L148), for which it looks like you can use [this method](https://github.com/bloomberg/clang-p2996/blob/p2996/clang/include/clang/AST/APValue.h#L653) to get the `ValueDecl`. Maybe we can just check if that `ValueDecl` is a `CXXMethodDecl`, and return that?

Feel free to ping me if more detail would help.

Reply:
Done. Let me know if it looks similar to what you expected